test/e2e/auth/audit.go

```diff
@@ -61,45 +71,123 @@ var _ = SIGDescribe("Advanced Audit [Feature:Audit]", func() {
+
+		config, err := framework.LoadConfig()
+		if err != nil {
+			framework.Failf("failed to load config: %v", err)
+		}
+		apiExtensionClient, err := clientset.NewForConfig(config)
+		if err != nil {
+			framework.Failf("failed to initialize apiExtensionClient: %v", err)
+		}
+		crd := testserver.NewRandomNameCustomResourceDefinition(apiextensionsv1beta1.ClusterScoped)
+		_, err = testserver.CreateNewCustomResourceDefinition(crd, apiExtensionClient, f.ClientPool)
+		testserver.DeleteCustomResourceDefinition(crd, apiExtensionClient)
+
+		expectedEvents := []auditEvent{}
+		expectedEvents = append(expectedEvents, commonExpectedEvents("pods", namespace, "create", pod.Name)...)
+		expectedEvents = append(expectedEvents, commonExpectedEvents("pods", namespace, "delete", pod.Name)...)
+		expectedEvents = append(expectedEvents, commonExpectedEvents("pods", namespace, "get", pod.Name)...)
+		expectedEvents = append(expectedEvents, commonExpectedEvents("pods", namespace, "list", pod.Name)...)
+
+		expectedEvents = append(expectedEvents, commonExpectedEvents("deployments", namespace, "create", d.Name)...)
+		expectedEvents = append(expectedEvents, commonExpectedEvents("deployments", namespace, "delete", d.Name)...)
+		expectedEvents = append(expectedEvents, commonExpectedEvents("deployments", namespace, "get", d.Name)...)
+		expectedEvents = append(expectedEvents, commonExpectedEvents("deployments", namespace, "list", d.Name)...)
+
+		expectedEvents = append(expectedEvents, commonExpectedEvents("secrets", namespace, "create", secret.Name)...)
+		expectedEvents = append(expectedEvents, commonExpectedEvents("secrets", namespace, "delete", secret.Name)...)
+		expectedEvents = append(expectedEvents, commonExpectedEvents("secrets", namespace, "get", secret.Name)...)
+		expectedEvents = append(expectedEvents, commonExpectedEvents("secrets", namespace, "list", secret.Name)...)
+
+		expectedEvents = append(expectedEvents, commonExpectedEvents("configmaps", namespace, "create", configMap.Name)...)
+		expectedEvents = append(expectedEvents, commonExpectedEvents("configmaps", namespace, "delete", configMap.Name)...)
+		expectedEvents = append(expectedEvents, commonExpectedEvents("configmaps", namespace, "get", configMap.Name)...)
+		expectedEvents = append(expectedEvents, commonExpectedEvents("configmaps", namespace, "list", configMap.Name)...)
+
+		expectedEvents = append(expectedEvents, watchExpectedEvents("pods", namespace)...)
+		expectedEvents = append(expectedEvents, watchExpectedEvents("deployments", namespace)...)
+		expectedEvents = append(expectedEvents, watchExpectedEvents("secrets", namespace)...)
+		expectedEvents = append(expectedEvents, watchExpectedEvents("configmaps", namespace)...)
+
+		expectedEvents = append(expectedEvents, customResourceEvents("create", *crd)...)
+		expectedEvents = append(expectedEvents, customResourceEvents("delete", *crd)...)
+
+		// Sleep for watches to timeout.
+		time.Sleep(3 * watchTimeout * time.Second)
```

Comment:
Why sleep? Sleeps are usually suspicious causing flakes under load.

Reply:
I think that we anyway need some kind of timeout here. Or we can check just for the first event. What do you prefer here?

Reply:
You can use channels for that, e.g. create a list of channels and sequentially wait for all of them to close

Reply:
Thanks for the idea! Done.